### PR TITLE
Bp/project visibility restricted

### PR DESCRIPTION
**Pull request by BartoszPrusak**
Internal fix for correct handling of the `VISIBILITY_RESTRICTED` API error.

### CODE REVIEW

File: src/neptune/internal/backends/swagger_client_wrapper.py [hunk 1]
```diff
@@ -28,66 +29,65 @@
...
+            ProjectPrivacyRestrictedException,
+            ProjectsLimitReached,
+        )
+
         try:
             body = response.json() or dict()
         except Exception:
             body = {}
 
+        error_processors: dict[str, Callable[[dict], ManagementOperationFailure]] = {
```
Comment (aniezurawski) [inline src/neptune/internal/backends/swagger_client_wrapper.py:53]: I don't think all these exception inherit from `ManagementOperationFailure`.

File: src/neptune/internal/backends/swagger_client_wrapper.py [hunk 2]
```diff
@@ -28,66 +29,65 @@
...
+            ),
+            "LIMIT_OF_PROJECTS_REACHED": lambda _: ProjectsLimitReached(),
+            "PROJECT_KEY_COLLISION": lambda response_body: ProjectKeyCollision(
+                key=response_body.get("key", "<unknown key>")
+            ),
+            "PROJECT_KEY_INVALID": lambda response_body: ProjectKeyInvalid(
+                key=response_body.get("key", "<unknown key>"),
+                reason=response_body.get("reason", "Unknown reason"),
+            ),
+            "PROJECT_NAME_COLLISION": lambda response_body: ProjectKeyCollision(
```
Comment (aniezurawski) [inline src/neptune/internal/backends/swagger_client_wrapper.py:72]: `ProjectNameCollision`

File: src/neptune/internal/backends/swagger_client_wrapper.py [hunk 3]
```diff
@@ -28,66 +29,65 @@
...
+                key=response_body.get("key", "<unknown key>")
+            ),
+            "PROJECT_NAME_INVALID": lambda response_body: ProjectNameCollision(
```
Comment (aniezurawski) [inline src/neptune/internal/backends/swagger_client_wrapper.py:75]: `ProjectNameInvalid`

File: src/neptune/internal/backends/swagger_client_wrapper.py [hunk 4]
```diff
@@ -28,66 +29,65 @@
...
+                name=response_body.get("name", "<unknown name>")
+            ),
+            "VISIBILITY_RESTRICTED": lambda response_body: ProjectPrivacyRestrictedException(
+                requested=response_body.get("requestedValue", "the selected"),
+                allowed=response_body.get("allowedValues"),
+            ),
+            "WORKSPACE_IN_READ_ONLY_MODE": lambda response_body: NeptuneLimitExceedException(
+                reason=response_body.get("title", "Unknown reason")
+            ),
+        }
```
Comment (aniezurawski) [inline src/neptune/internal/backends/swagger_client_wrapper.py:85]: @Raalsky @BartoszPrusak how do you feel about testing these not-happy paths e2e? Do you think it's worth the effort?

**Review by aniezurawski — CHANGES_REQUESTED** [review]

File: src/neptune/internal/backends/swagger_client_wrapper.py (hunk 4 above)
Comment (Raalsky) [reply src/neptune/internal/backends/swagger_client_wrapper.py:85]: In the long term, I would like to add some kind of integration tests that will mimic the possible outcomes of a backend API (mocking responses of `requests`). That would be a much better place for such not-happy paths and e2e could be a bare minimum. Currently -> It's ok but there is no big need for it.

File: src/neptune/internal/backends/swagger_client_wrapper.py (hunk 1 above)
Comment (BartoszPrusak) [reply src/neptune/internal/backends/swagger_client_wrapper.py]: Right, I forgot about `NeptuneException`s.
Fixed.

File: src/neptune/internal/backends/swagger_client_wrapper.py (hunk 3 above)
Comment (BartoszPrusak) [reply src/neptune/internal/backends/swagger_client_wrapper.py]: Fixed

File: src/neptune/management/exceptions.py [hunk 5]
```diff
@@ -178,3 +178,16 @@ class WorkspaceOrUserNotFound(ManagementOperationFailure):
 class UserAlreadyInvited(ManagementOperationFailure):
     code = 24
     description = "User '{user}' has already been invited to the workspace '{workspace}'."
+
+
+class ProjectPrivacyRestrictedException(ManagementOperationFailure):
+    code = 25
+    description = "Cannot set {requested} visibility for project. {followup}"
```
[the remark this answers is not in the document]
Comment (BartoszPrusak) [reply src/neptune/management/exceptions.py]: I made a new commit in this MR with simplified message. Hit me up if you need to know more :)

File: src/neptune/internal/backends/swagger_client_wrapper.py (hunk 4 above)
Comment (BartoszPrusak) [reply src/neptune/internal/backends/swagger_client_wrapper.py:85]: I added an e2e test for the error message I introduced, but don't plan on adding more.
I talked to Raalsky and the non-happy paths will be tested in a different environment.

**Review by aniezurawski — APPROVED** [review]

**Pull request merged**